course/models.py

```diff
@@ -1704,11 +1716,55 @@ def consume(self, iterable, set_is_superseded=False):
             gchange.is_superseded = False
             self._consume_grade_change(gchange, set_is_superseded)
 
-        valid_grade_changes = sorted(
-                (gchange
-                for gchange in self.attempt_id_to_gchange.values()
-                if gchange.percentage() is not None),
-                key=lambda gchange: gchange.grade_time)
+        def get_valid_grade_changes():
+            # type: () -> List[GradeChange]
+            gchanges = [gchange for gchange in self.attempt_id_to_gchange.values()
+                       if gchange.percentage() is not None]
+            gchanges_with_session = sorted([
+                gchange for gchange in gchanges
+                if gchange.flow_session is not None],
+                key=lambda x: (x.flow_session.previous_completion_time, x.pk))
+            n_completed_sessions = len(gchanges_with_session)
+            if n_completed_sessions <= 1:
+                return sorted(
+                    gchanges, key=lambda x: x.flow_session.previous_completion_time)
+
+            flags = [0] * n_completed_sessions
+            flags[0] = -1
+            flags[-1] = 1
+            gchange_flag_tuple = list(zip(gchanges_with_session, flags))
+
+            gchanges_with_out_session = sorted([
+                gchange for gchange in gchanges
+                if gchange.flow_session is None],
+                key=lambda x: x.grade_time
+            )
+
+            time_pre_completed_sessions = min(gchanges_with_session[0].grade_time,
+                                              gchanges_with_session[-1].grade_time)
+            time_post_completed_sessions = max(gchanges_with_session[0].grade_time,
+                                               gchanges_with_session[-1].grade_time)
+
+            for gchange in gchanges_with_out_session:
+                if gchange.grade_time < time_pre_completed_sessions:
+                    flag = -2
+                elif gchange.grade_time == time_pre_completed_sessions:
+                    flag = -1
+                elif gchange.grade_time == time_post_completed_sessions:
+                    flag = 1
+                elif gchange.grade_time > time_post_completed_sessions:
+                    flag = 2
+                else:
+                    flag = 0
+                gchange_flag_tuple.append((gchange, flag))
```

Comment:
I added an axes to compare GC objects. We don't need full sort, just return a list with a correct first and last GC is ok.